script/plugins/InputRules/inputrules.js

```diff
@@ -0,0 +1,42 @@
+import { inputRules, InputRule } from 'prosemirror-inputrules';
+import { TextSelection } from 'prosemirror-state';
+import SmileyConf from '../../custom/SmileyConf';
+
+/**
+ * Smiley input rule
+ *
+ * @returns {InputRule}
+ */
+export function smileyRule() {
+    return new InputRule(SmileyConf.getRegex(), ((state, match) => {
+        const { tr } = state;
+        const syntax = match[0];
+
+        // get icon corresponding to the captured group
+        let group = 0;
+        for (let i = 0; i < match.length; i += 1) {
+            if (i > 0 && match[i] === syntax) {
+                group = i;
+                break;
+            }
+        }
+        const { icon } = SmileyConf.getSmileys()[group - 1];
```

Comment:
This loop seems hard to grasp to me. What is inside of `match`?

To me it would seem easier to understand if we would have a map syntax->icon. We could then use that map with the syntax from line 13 to get the icon. (I'm speculating about the internal structure of `match` here)

Reply:
Thanks @micgro42 , an excellent point about random substitution. Currently it's not handled very well and probably buggy from the get-go The trailing space is a must, but it  makes sense to allow lines to start with a smiley, e.g. FIXME. I have to give it some more thought.

As to your other remark about syntax->icon map: that was my initial approach, but it's very tricky to work with the smiley syntax as keys. Hence the workaround:

Looping through the array of matched groups returned by `RegExp.exec()` tells me which group syntax was matched so I can fetch the right item from conf  `[ { syntax: 'X', icon: 'Y' }, { syntax: 'Z', icon: 'Q' } ]`

Maybe I can find a more transparent solution.

Reply:
Following the review by @micgro42 I made the following changes: 

1. Icons are matched to smileys using a simple map.
2. The regex pattern is almost the same as in Dokuwiki parser (except for a lookbehind, which is not supported by most browsers https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp#Browser_compatibility).